### PR TITLE
LPS-177798: As a developer I can upsert an object entry and it's related objects in a single request custom-system

### PR DESCRIPTION
**Pull request by sergiojimcos**
**What is new?**
- Add two new implementations to manage the parsering with System Object Entries, they are going to manage `Map<String, Object>`.
- Make some changes in `ObjectDefinitionDeployer` to register those two implementations.
- Add `upsertBaseModel` in `SystemObjectDefinitionMetada` to upsert a System Object Entry and implement this method in all implementations. 
- Update `DefaultObjectEntryManager` to upsert System Object Entries
- Create some tests cases that upsert Custom Object Entries with System Object Entries

**Note**: To test this functionality, it is necessary to have these feature flags activated:
- `feature.flag.LPS-153117 = true`

### Example
Let's see an example. Imagine that we have a `M-M` relationship with Subjects and their Students (System Object Definition) in this way:

- `Student` is a Custom Object. Has a many-to-many relationship (`studentSubject`) with `Subject which is a User`. To create a new subject and their related students, we can do this:

```bash
curl --request POST 'http://localhost:8080/o/c/subjects/' \
--header 'Content-Type: application/json' \
--data-raw '{"name": "English",
  "externalReferenceCode": "English",
  "studentSubject": [{
  
  "additionalName": "john",
  "alternateName": "johndoe",
  "emailAddress": "johndoe@liferay.com",
  "honorificPrefix": "john",
  "honorificSuffix": "doe",
  "password": "password"
  }]
}'
```
This request is going to create a new subject `English` and their related students (Users): `John Doe` and link them because they have a `M-M` relationship (`studentSubject`). 

### How to deploy it

Deploy the following modules:
- Deploy `account-api` and `account-service`.
- Deploy `commerce-api` and `commerce-service`.
- Deploy `object-res-api `and `object-rest-impl`.
- Deploy `object-api` and `object-service`

**JIRA Ticket:**
https://issues.liferay.com/browse/LPS-172384
https://issues.liferay.com/browse/LPS-172386

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java [hunk 1]
```diff
@@ -190,9 +193,41 @@ public synchronized void undeploy(ObjectDefinition objectDefinition) {
...
+			boolean unregisterSystemObjectDefinition = true;
+
 			JaxRsApplicationDescriptor jaxRsApplicationDescriptor =
 				systemObjectDefinitionMetadata.getJaxRsApplicationDescriptor();
 
+			Map<Long, List<ServiceRegistration<?>>> serviceRegistrationsMap =
+				_scopedServiceRegistrationsMap.get(
+					jaxRsApplicationDescriptor.getRESTContextPath());
+
+			if (serviceRegistrationsMap != null) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java:205]: I think this code is duplicated. Could we refactor this to avoid duplicating it?

**Review by ccorreagg — CHANGES_REQUESTED** [review]
See my comments

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryRelatedObjectsResourceTest.java [hunk 2]
```diff
@@ -815,7 +1045,10 @@ private int _invokeGetHttpCode(String location) throws Exception {
 		return response.getResponseCode();
 	}
 
-	private UserAccount _randomUserAccount() throws Exception {
+	private UserAccount _randomUserAccount(
+			String userExternalReferenceCode, String userEmailAdress)
+		throws Exception {
```
Comment (LuismiBarcos) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryRelatedObjectsResourceTest.java:886]: 🤔 🤔 🤔 🤔 
If you have to provide some values it is not random anymore, right? 😅 
AFAIK, a pattern that we use in other APIs is use the `randomEntity` method and after the creation setting the values that we want to have. For example:
```
UserAccount userAccount = new _randomUserAccount();
userAccount.setEmailAddress(yourEmailAddress)
userAccount.setExternalReferenceCode(yourExternalReferenceCode)
...
```

**Review by LuismiBarcos — COMMENTED** [review]
Just a picky comment from my side 😁 
Review Carlos's comments

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryRelatedObjectsResourceTest.java (hunk 2 above)
Comment (sergiojimcos) [reply modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryRelatedObjectsResourceTest.java:886]: Make sense @LuismiBarcos!!

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java [hunk 3]
```diff
@@ -190,9 +193,41 @@ public synchronized void undeploy(ObjectDefinition objectDefinition) {
...
+			Map<Long, List<ServiceRegistration<?>>> serviceRegistrationsMap =
+				_getServiceRegistrationsMap(
+					jaxRsApplicationDescriptor.getRESTContextPath());
+
+			if (serviceRegistrationsMap != null) {
+				List<ServiceRegistration<?>> serviceRegistrations =
+					serviceRegistrationsMap.remove(
+						objectDefinition.getCompanyId());
+
+				if (serviceRegistrations != null) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java:210]: I think this code is duplicated. Could we refactor this to avoid duplicating it?

File: modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java [hunk 4]
```diff
@@ -72,6 +73,24 @@ public String getTitleObjectFieldName() {
...
+	public long upsertBaseModel(
+			String externalReferenceCode, long companyId, User user,
+			Map<String, Object> values)
+		throws Exception {
+
+		BaseModel<?> baseModel = fetchBaseModelByExternalReferenceCode(
+			externalReferenceCode, companyId);
+
+		if (baseModel != null) {
+			updateBaseModel((Long)baseModel.getPrimaryKeyObj(), user, values);
```
Comment (ccorreagg) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java:86]: Maybe we could do the casting before the method and then, return that casted element:

```
long primaryKey = (long)baseModel.getPrimaryKeyObj();

updateBaseModel(primaryKey, user, values);

return primaryKey;
```

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Review my comments!

File: modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java [hunk 5]
```diff
@@ -72,6 +73,26 @@ public String getTitleObjectFieldName() {
...
+	public long upsertBaseModel(
+			String externalReferenceCode, long companyId, User user,
+			Map<String, Object> values)
+		throws Exception {
+
+		BaseModel<?> baseModel = fetchBaseModelByExternalReferenceCode(
+			externalReferenceCode, companyId);
+
+		if (baseModel != null) {
+			long primaryKey = (Long)baseModel.getPrimaryKeyObj();
```
Comment (ccorreagg) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java:86]: Maybe better cast to long instead:

```
long primaryKey = (long)baseModel.getPrimaryKeyObj();
```

Comment (sergiojimcos) [reply modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java:86]: Ups sorry, makes sense! Better to cast it using the primitive instead of the Object

**Review by ccorreagg — COMMENTED** [review]
Just a comment!

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java [hunk 6]
```diff
@@ -727,6 +634,147 @@ private void _initSystemObjectDefinition(
 			});
 	}
 
+	private boolean _shouldUnregisterApplication(String restContextPath) {
+		List<String> companyIds = _restContextPathCompanyIds.get(
+			restContextPath);
+
+		if ((companyIds != null) && !companyIds.isEmpty()) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java:641]: Maybe more clear:

```
if (ListUtil.isNotEmpty(companyIds)) {
```

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java [hunk 7]
```diff
@@ -727,6 +636,151 @@ private void _initSystemObjectDefinition(
 			});
 	}
 
+	private boolean _shouldUnregisterApplication(String restContextPath) {
+		List<String> companyIds = _restContextPathCompanyIds.get(
+			restContextPath);
+
+		if ((companyIds != null) && ListUtil.isNotEmpty(companyIds)) {
```
Comment (LuismiBarcos) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java:643]: IntelliJ warnings are useful. Seems like `isNotEmpty` also check for nulls, so, the first check is not necessary.
image

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java [hunk 8]
```diff
@@ -727,6 +636,151 @@ private void _initSystemObjectDefinition(
...
+
+		_undeployObjectDefinitionsMap(companyId, restContextPath);
+		_undeployRestContextPathCompanyIds(companyId, restContextPath);
+		_undeployScopedServiceRegistrationsMap(restContextPath, companyId);
+
+		boolean unregisterApplication = _shouldUnregisterApplication(
+			restContextPath);
+
+		if (unregisterApplication) {
+			_unregisterApplication(unregisterApplication, restContextPath);
```
Comment (LuismiBarcos) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/deployer/ObjectDefinitionDeployerImpl.java:673]: If we want to check to unregister the application before calling the method (I am ok with that, maybe is clearer indeed) I think it is not necessary to pass the boolean `unregisterApplication`. Why? Because here the unregisterApplication will be always true.

Inside the `_unregisterApplication` method, if you check the use of unregisterApplication you will see that it only use here
```
if (!shouldUnregisterApplication) {
	return;
}
```
If unregisterApplication is always true, this `return` statement will never be reached, so, the if is not necessary, which makes unuseful the unregisterApplication argument.

So, to sum up, we can check if it is necessary to unregister the application before calling the method, but in that case, the variable to check if should unregister the application is not necessary inside the `_unregisterApplication` method 🤷 

NOTE: In fact, in Carlos's suggestion about checking this before the unregister method, he proposed to delete the argument in the method
image

**Review by LuismiBarcos — COMMENTED** [review]
I left some picky comments, you know 😅

**Comment by ccorreagg** [comment]
Followed-up PR (including an extra commit): https://github.com/liferay-headless/liferay-portal/pull/1145
